### PR TITLE
Dt docs for 2.3 part 3, and some test cases as well

### LINKED ISSUE #25469: Fix devicetree documentation for new API
The devicetree documentation contains errors and is missing descriptions due to late-breaking semantic changes which had to go in before code freeze. This issue tracks the necessary work I'm aware of. Release notes are tracked separately in #22637.

- [x] DTC_OVERLAY_FILE description says colon-separated, that's wrong (https://github.com/zephyrproject-rtos/zephyr/pull/25470)
- [x] centralize documentation of chosen zephyr nodes in a non-legacy file, provide a reference to them from the intro (https://github.com/zephyrproject-rtos/zephyr/pull/25470)
- [x] review doxygen docstrings and correct errors for generic APIs (https://github.com/zephyrproject-rtos/zephyr/pull/25470)
- [x] add introductory text to each section in the API reference (https://github.com/zephyrproject-rtos/zephyr/pull/25470)
- [x] review doxygen docstrings and correct errors for instance APIs (#25633)
- [x] review doxygen docstrings and correct errors for hw specific APIs (#25633)
- [x] address outstanding comments from #25009 (#25633)
- [x] review user guides and correct errors (#25633)
- [ ] write a migration guide for moving to the new API
- [ ] check if zephyr v2.2 restriction in https://docs.zephyrproject.org/latest/guides/dts/bindings.html#bindings-file-syntax is still an issue

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Along with https://github.com/zephyrproject-rtos/zephyr/pull/25633 and https://github.com/zephyrproject-rtos/zephyr/pull/25470 (already merged), this PR:

Fixes: #25469

By:

- writing a migration guide for moving to the new API
- checking if zephyr v2.2 restriction in https://docs.zephyrproject.org/latest/guides/dts/bindings.html#bindings-file-syntax is still an issue (formerly, only 2 levels of child bindings were permitted, but now we can apparently keep recursing to our heart's delight)

I wanted to make sure the migration guide doesn't regress and the bindings restriction is indeed lifted in perpetuity, so I added test cases as well. I did have to touch gen_legacy_defines.py to make those tests cases not throw warnings about deprecated macros.

With this PR, I also give my sign-off to the DT release notes for #22637.

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
@hakonfam you originally proposed the migration guide, so any feedback on this effort would be welcome. Thanks!

File: samples/hello_world/app.overlay [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/ {
```
Comment (pabigot) [inline samples/hello_world/app.overlay:1]: Why is this here?

Comment (galak) [reply samples/hello_world/app.overlay:1]: ditto

**Review by galak — CHANGES_REQUESTED** [review]
Looks good, beyond addition of samples/hello_world/app.overlay

File: samples/hello_world/app.overlay (hunk 1 above)
Comment (mbolivar-nordic) [reply samples/hello_world/app.overlay]: sorry about that

**Review by galak — APPROVED** [review]
Not sure if you can add a license/copyright header to migration-example.inc to make scancode happy.

**Review by pabigot — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
> Not sure if you can add a license/copyright header to migration-example.inc to make scancode happy.

I tried that with the C file but I couldn't get the presentation to come out right using `:start-after:`; there was an extra newlines at the start of the file. I'm just going to inline the content even though that's not as nice.

**Review by carlescufi — APPROVED** [review]
